src/main/java/com/conveyal/r5/analyst/progress/Task.java

```diff
@@ -50,66 +45,63 @@
 
     private int logFrequency = 0;
 
-    public String description;
+    public int totalWorkUnits = 1;
 
-    public int totalWorkUnits;
+    public int currentWorkUnit = 0;
 
-    public int currentWorkUnit;
+    public final Map<String, String> tags = new HashMap<>();
 
-    public static enum State {
-        QUEUED, ACTIVE, DONE, ERROR
+    static class LogEntry {
+        public final long time = Instant.now().toEpochMilli();
+        public final Level level;
+        public final String message;
+        public LogEntry (String message, Level level) {
+            this.message = message;
+            this.level = level;
+        }
     }
+    public final List<LogEntry> log = new ArrayList<>();
 
-    public State state;
+    public enum State {
+        QUEUED, ACTIVE, DONE, ERROR
+    }
 
-    /** To be on the safe side all tasks are considered heavyweight unless we explicitly set them to be lightweight. */
-    private boolean isHeavy = true;
```

Comment:
Isn't the designation of "heavy" a characteristic of the scheduler, not the task? The scheduler could have a `medium` thread also right? Or a `super heavy`.

Reply:
As described in the Javadoc, the two executors are intended as a rough approximation of a work-stealing pool, or more practically speaking, to make sure quick tasks don't get stuck behind a pile of very slow ones. I don't know what you mean by "heavy" being a characteristic of the scheduler. The scheduler wants to categorize the tasks, so the tasks declare whether they are usually light or heavy, to a rough order of magnitude ("sure to finish in 1-2 seconds" is distinct from "may take 5 minutes"). So "light" and "heavy" describe the tasks, and are also the names given in the scheduling implementation to two executors that handle tasks in those two categories. Because we created all pieces of the puzzle, we can conveniently use this same two-valued categorization of tasks inside the TaskScheduler.

I suppose yes, if you can have two executors you could have any number of them, and the units used to express task "weight" would not have to be the same as those used describe these N executors. But I don't think that affects the underlying idea here: we construct a task and declare all its characteristics. The task is submitted to the scheduler and handled based on its declared characteristics. Conceivably those characteristics could be provided some other way like additional interface methods on TaskAction, which I considered, but then TaskAction wouldn't be a single-function interface and could not be constructed via anonymous functions, which is sometimes convenient.

Reply:
Maybe another source of confusion here is that the same class (Task) is serving as an internal class and a REST API model class. So some characteristics may be relevant to the internal object (its own estimate of how slow it is) but are just noise to the end user. We can silence some of those fields with serialization annotations. This is definitely a shortcut (to avoid creating tons of separate API model classes) but it's a shortcut I'm generally comfortable with on simple classes without much logic.

Reply:
I mentioned this more here: https://github.com/conveyal/r5/pull/716#issuecomment-811433523

> An additional question I have, should every Task be "heavy"? If it's not, it shouldn't be a task.
>
> That would simplify the scheduler more. The EventBus could just have its own light queue. Although with the addition of Tasks/TaskScheduler we may want to make all tasks either synchronous and skip the scheduling or asynchronous and queue them all.

Note: I meant to say "we may want to make all **events** either synchronous and skip the scheduling or asynchronous and queue them all."

I think another source of confusion is the placement of the `Task.isHeavy` flag in the `r5` package while the only usage is in the `TaskScheduler` in the `analysis` package.